[MLv2] Boolean Picker Conversion

iethree opened the pull request:
> **Warning** 
> MLv2 blows up if you select any date fields

closes https://github.com/metabase/metabase/issues/33539

### Description

- pulls in changes from Backend PRs: 
  - https://github.com/metabase/metabase/pull/33623
- Adds `FilterContext` in FilterPopover to allow MLv1 and MLv2 subcomponents to live alongside each other in peace
- fixes a bug in FilterPopover that broke `empty` and `not-empty` boolean filters
- Updates BooleanPickerRadio to use MLv2 filter manipulattion functions
- update BooleanPicker to use `metabase/ui` radio component
- updates tests

![Screen Shot 2023-08-31 at 1 51 04 PM](https://github.com/metabase/metabase/assets/30528226/102bbd70-1ac2-4ceb-a10d-2cb67d26b103)

### How to verify

see that you can:

- [ ] add, edit, and remove filters for a boolean column in the notebook editor
- [ ] add, edit, and remove filters for a boolean column using column headers
- [ ] add, edit, and remove post-aggregation filters for a boolean column in the notebook editor
- [ ] add, edit, and remove post-aggregation filters for  a boolean column using column headers
- [ ] see that filters added in the notebook appear in the bulk filter modal (which is still on MLv1)
- [ ] see that filters added in the bulk filter modal, appear in the notebook

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

iethree committed:
lift legacy-field-ref from #33623

iethree committed:
filters compatibility

iethree committed:
convert BooleanPicker to MLv2

iethree commented on frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx:
```diff
@@ -242,83 +246,103 @@ export function FilterPopover({
 
   const filterOperator = filter.operator();
-  const hasPicker = filterOperator && filterOperator.fields.length > 0;
+  const hasPicker =
+    (filterOperator && filterOperator.fields.length > 0) || isBoolean(field);
```
this caused `empty` and `not-empty` boolean filters to make the picker disappear 🤦

kulyk replied:
I think it might be worth adding a comment explaining the thing
It'd be easy to forget why we had to treat boolean fields differently

iethree commented on frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx:
```diff
@@ -242,83 +246,103 @@ export function FilterPopover({
-  const hasPicker = filterOperator && filterOperator.fields.length > 0;
+  const hasPicker =
+    (filterOperator && filterOperator.fields.length > 0) || isBoolean(field);
+  const stageIndex = -1; // this will not always be true
```
this works... not sure why.

kulyk replied:
Do you mean the `-1` value? It's a shortcut for "the last query stage". But yeah, we shouldn't be using it here

iethree commented on frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopoverPicker.tsx:
```diff
@@ -63,9 +63,8 @@ export class FilterPopoverPicker extends Component<Props> {
       />
     ) : field?.isBoolean() ? (
-      <BooleanPicker
+      <BooleanPickerRadio
```
just cleaning up default exports here

iethree commented on frontend/src/metabase/query_builder/components/filters/pickers/BooleanPicker/BooleanPicker.unit.spec.tsx:
```diff
@@ -28,6 +47,20 @@ describe("BooleanPicker", () => {
+      }),
+    ],
+    tables: [
+      createMockTable({
```
MLv2 needs table metadata for this test to work

iethree commented on frontend/src/metabase/query_builder/components/filters/pickers/BooleanPicker/BooleanPickerRadio.tsx:
this basically needed a full component rewrite.

iethree commented on frontend/src/metabase/query_builder/components/filters/pickers/BooleanPicker/BooleanPickerRadio.tsx:
```diff
@@ -1,48 +1,81 @@
-        value={value}
-        onChange={updateFilter}
-      />
+      <Radio.Group value={displayValue} onChange={updateFilter}>
```
swapped out the radio input for a non-deprecated component.

iethree commented on frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx:
```diff
@@ -242,83 +246,103 @@ export function FilterPopover({
+    (filterOperator && filterOperator.fields.length > 0) || isBoolean(field);
+  const stageIndex = -1; // this will not always be true
+
+  const {
+    filterClause,
+    query: mlv2Query,
+    column,
+  } = filter.getMlv2FilterClause({ stageIndex });
```
pretty sure this is what blows up if you give it a date column
```
Don't know how to set options for [nil {} [:field {:lib/uuid "8841e4ea-26c5-4b04-9137-bca630130097", :base-type :type/Date, :effective-type :type/Date} 2169]]
```

iethree commented on frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx:
```diff
@@ -242,83 +246,110 @@ export function FilterPopover({
+    filterClause,
+    query: mlv2Query,
+    column,
+    // TODO: don't blow up trying to make an MLv2 Date filter
+  } = isDate(dimension)
```
😬  this throws the error:

```
Don't know how to set options for [nil {} [:field {:lib/uuid "b9d9d3d5-e1d8-45b9-aeaa-2e7760327ceb", :base-type :type/Date, :effective-type :type/Date} 2169] -1 "week" #js {:include_current false}]
```

kulyk commented on frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx:
```diff
@@ -242,83 +246,110 @@ export function FilterPopover({
-          disableChangingDimension={!showFieldPicker}
-          supportsExpressions={supportsExpressions}
+      <ErrorBoundary>
+        <FilterContext.Provider
+          value={{
+            filter: filterClause,
+            query: mlv2Query,
+            legacyQuery: query,
+            column: column,
+            stageIndex,
+          }}
```
I think it's worth introducing a `FilterContextProvider` component that'd take an MLv1 filter and do all the conversation heavy lifting internally. We should also be able to guard ourselves from null-ish queries in there

iethree replied:
yeah, I think using context was just slightly too clever here. take a look at the version without context and see what you think.

kulyk commented on frontend/src/metabase-lib/references.ts:
```diff
@@ -106,6 +106,7 @@ export const getDimensionReferenceWithoutOptions = (
 export const BASE_DIMENSION_REFERENCE_OMIT_OPTIONS = [
   "temporal-unit",
   "binning",
+  "base-type",
 ];
```
I'm worried this might blow up. Why do we need it?

iethree replied:
yeah, I was a little unsure about this, but this is needed to make MLv1 and MLv2 columns match up, without it, a filter created using a popover will have a column like: `['field', 99, { "base-type": boolean } ]` and the filterPopover will be looking for `['field', 99, null ]` and the `isSameBaseDimension()` check will be false, so they won't match up. 😢

iethree replied:
I think this should be safe when it's used for comparison, we're not actually mutating any queries that we send to be executed.

kulyk commented on frontend/src/metabase-lib/queries/structured/Filter.ts:
```diff
@@ -400,4 +403,8 @@ export default class Filter extends MBQLClause {
     return operator && operator.name === operatorName;
   }
+
+  getMlv2FilterClause({ stageIndex = -1 }) {
+    return getMlv2FilterClause(this, stageIndex);
+  }
```
Instead of passing the `stageIndex`, we can actually calculate it from the filter's `query` object:

```ts
// Adding a method
class StructuredQuery extends AtomicQuery {
  getQueryStageIndex() {
    return this.queries().length - 1;
  }
}

// Using the method
class Filter extends MBQLClause {
  getMlv2FilterClause() {
    const query = this._query;
    const mlv2Query = this._query.rootQuery().getMLv2Query()
    const stageIndex = query.getQueryStageIndex();
  }
}
```

kulyk commented on frontend/src/metabase-lib/compat/compat.ts:
```diff
@@ -0,0 +1,96 @@
+};
+
+export const getMlv2FilterClause = (filter: LegacyFilter, stageIndex = -1) => {
+  const mlv2Query = filter.query().question()._getMLv2Query();
```
Not sure if `filter.query().question()` will give you the top-level query. I think `filter.query().rootQuery().getMLv2Query()` should do the thing

kulyk replied:
Also, I think we should be calling an MLv2 query just `query` and prefix MLv1 stuff with `legacy` (`legacyQuery`, `legacyFilter`, etc.)

iethree replied:
.getMLv2Query is private on query so we have to 
`const query = filter.query().rootQuery().question()._getMLv2Query();` 
😜

kulyk commented on frontend/src/metabase-lib/compat/compat.ts:
```diff
@@ -0,0 +1,96 @@
+export const getMlv2FilterClause = (filter: LegacyFilter, stageIndex = -1) => {
+  const mlv2Query = filter.query().question()._getMLv2Query();
+
+  const appliedFilterClause = ML.findFilterForLegacyFilter(
+    mlv2Query,
+    stageIndex,
+    [...(filter as unknown as FieldFilter)],
+  );
```
We should be able to replace `[...(filter as unknown as FieldFilter)]` with something like that:

```ts
class Filter extends MBQLClause {
  raw(): FieldFilter {
    return [...this]
  }
}

ML.findFilterForLegacyFilter(
    mlv2Query,
    stageIndex,
    filter.raw(),
  );
```

kulyk commented on frontend/src/metabase-lib/compat/compat.ts:
```diff
@@ -0,0 +1,96 @@
+
+  const column = appliedFilterClause
+    ? ML.filterParts(mlv2Query, stageIndex, appliedFilterClause).column
+    : ML.findFilterableColumnForLegacyRef(mlv2Query, stageIndex, filter[1]);
```
I think we should better use a method to get the filter field reference:

```ts
filter.dimension().mbql()
```

kulyk commented on frontend/src/metabase-lib/compat/compat.ts:
```diff
@@ -0,0 +1,96 @@
+    stageIndex,
+  });
+
+  const unappliedFilterClause = ML.filterClause(
+    operatorsMap[filter[0]],
+    column as ML.ColumnMetadata,
+    ...filter.slice(2),
+  );
```
Let's use filter methods to access filter properties instead of taking directly from MBQL:

```ts
const operator = operatorsMap[filter.operatorName()]
const args = filter.arguments()

ML.filterClause(operator, column, args);
```

kulyk replied:
Also, why do we need to do `column as ML.ColumnMetadata` ?

kulyk commented on frontend/src/metabase-lib/compat/compat.ts:
```diff
@@ -0,0 +1,96 @@
+  legacyQuery: StructuredQuery,
+  filterClause: ML.FilterClause,
+): LegacyFilter {
+  const { operator, column, options, args } = ML.filterParts(
+    query,
+    -1,
+    filterClause,
+  );
```
Please don't hardcode the `stageIndex`, let's make it an argument

kulyk commented on frontend/src/metabase-lib/compat/compat.ts:
```diff
@@ -0,0 +1,96 @@
+import type StructuredQuery from "metabase-lib/queries/StructuredQuery";
+import LegacyFilter from "metabase-lib/queries/structured/Filter";
+
+export const getOperatorsMap = ({
+  query,
+  column,
+  stageIndex = -1,
+}: {
+  query?: ML.Query;
+  column?: ML.ColumnWithOperators;
+  stageIndex?: number;
+}) => {
```
WDYT about making the signature consistent with the rest of MLv2 functions? Most functions usually expect a `query` and a `stageIndex` as first two arguments

kulyk reviewed (changes requested):
Looks good! Left some questions and comments, I think my key takeaways are:

* we should fix how we work with queries and stage index in `compat` files
* we should have a `FilterContextProvider` component and a `useFilterContext` hook to encapsulate some logic and do some type guarding

ranquild commented on frontend/src/metabase/query_builder/components/filters/pickers/BooleanPicker/BooleanPickerRadio.tsx:
```diff
@@ -1,48 +1,81 @@
-    } else if (typeof value === "string") {
-      onFilterChange(filter.setOperator(value));
-    }
+  const { filter, query, legacyQuery, column, stageIndex } =
+    useContext(FilterContext);
+
+  const {
+    args: [value],
+    operator,
+  } = filter
+    ? ML.filterParts(query as ML.Query, stageIndex, filter)
+    : { args: [], operator: undefined };
```
Some filters have 2 columns; some others use expressions and not columns

ranquild reviewed (changes requested):
Posted feedback here https://metaboat.slack.com/archives/C04CYTEL9N2/p1693828897858269. I'd like to discuss this before going further with this PR.

iethree commented on frontend/src/metabase/query_builder/components/filters/pickers/BooleanPicker/BooleanPickerRadio.tsx:
```diff
@@ -1,48 +1,81 @@
+    [column, query, stageIndex],
+  );
+
+  if (!query || !legacyQuery || !column) {
+    return null;
+  }
```
yeah, I think we may also be able to tweak the context setup to make this unnecessary.

iethree commented on frontend/src/metabase-lib/filter.ts:
```diff
@@ -28,9 +29,13 @@ export function filterableColumnOperators(
 export function filterClause(
   filterOperator: FilterOperator,
   column: ColumnMetadata,
-  ...args: ExpressionArg[]
+  ...args: (ExpressionArg | undefined)[]
 ): FilterClause {
-  return ML.filter_clause(filterOperator, column, ...args);
+  if (args.some(arg => arg !== undefined)) {
+    return ML.filter_clause(filterOperator, column, ...args);
+  }
+
+  return ML.filter_clause(filterOperator, column);
```
I think I found a nicer way to do this. 👍

iethree commented on frontend/src/metabase/common/context/FilterContext/FilterContext.tsx:
```diff
@@ -0,0 +1,20 @@
+  stageIndex: number;
+}
+
+export const FilterContext = createContext<FilterContextType>({
+  filter: undefined,
+  query: undefined,
+  legacyQuery: undefined,
+  column: undefined,
+  stageIndex: -1,
+});
```
I think I've come around to not using the context and just converting to MLv2 in the component.

iethree committed:
make booleanPicker better

iethree commented:
a few updates:
1. Removed FilterContext, things are simpler just converting in each component.
2. Updated filterClause to take args as an array and have a separate options parameter.
3. Updated the booleanPicker to be more option-centric, rather than MBQL-centric
4. lots of little tweaks to conversion logic to make it more robust (thanks @kulyk !)

kulyk commented on frontend/src/metabase/query_builder/components/filters/pickers/BooleanPicker/BooleanPickerRadio.tsx:
```diff
@@ -1,48 +1,97 @@
+
+  const filterOptions = useMemo(
+    () => [
+      {
+        name: t`True`,
+        value: "true",
+        filter: ML.filterClause(operatorsMap["="], column, [true]),
+      },
```
Oh, that's nice ✨

kulyk replied:
Wondering if we could use a clause as a value here, so we don't need to look it up later?

iethree replied:
the new radio component doesn't allow object values. At first I was thinking about extending the types, but I think the lookup is just as good, plus, matching objects to figure out which radio item is selected would be expensive and possibly error-prone.

kulyk commented on frontend/src/metabase/query_builder/components/filters/pickers/BooleanPicker/BooleanPicker.unit.spec.tsx:
```diff
@@ -1,33 +1,42 @@
-  return render(
-    <BooleanPicker filter={filter} onFilterChange={mockOnFilterChange} />,
+
+  return renderWithProviders(
+    <BooleanPickerRadio filter={filter} onFilterChange={mockOnFilterChange} />,
   );
```
Why do we need `renderWithProviders` now?

kulyk reviewed (commented):
Nice work! Left a few more comments

iethree commented on frontend/src/metabase-lib/compat/compat.ts:
```diff
@@ -0,0 +1,99 @@
+  const query = filter.query().rootQuery().question()._getMLv2Query();
+  const stageIndex = filter.query().getStageIndex();
+
+  const appliedFilterClause = ML.findFilterForLegacyFilter(
+    query,
+    stageIndex,
+    filter.raw() as FieldFilter,
+  );
```
oh, I missed the typing on the `raw` method 👍

iethree commented:
I think we're gonna revisit the MLv2 API before moving forward here.
